invenio_rdm_records/resources/deserializers/rocrate/__init__.py

```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2022 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""RO-Crate deserializer."""
+
+from flask_resources import JSONDeserializer
+
+from .schema import ROCrateSchema
+
+
+class ROCrateJSONDeserializer(JSONDeserializer):
+    """RO-Crate metadata deserializer."""
+
+    def __init__(self, schema=ROCrateSchema):
+        """."""
+        self.schema = schema
+
+    def deserialize(self, data):
+        """Deserialize the RO-Crate payload."""
+        data = super().deserialize(data)
+        graph = data.get("@graph")
+        if not graph:
+            # TODO: Raise exception?
+            pass
+
+        dataset = self._dereference(graph)
+        metadata = self.schema().load(dataset)
+        return {"metadata": metadata}
+
+    @staticmethod
+    def _dereference(data: list):
+        entities = {}
+        for item in data:
+            if "@id" in item:
+                entities[item["@id"]] = item
+            item.pop("@reverse", None)
+
+        dataset = entities.pop("./")
+
+        def _walk(node):
+            if isinstance(node, dict):
+                for key in list(node.keys()):
+                    value = node[key]
+                    if key == "@id" and value in entities:
+                        node.update(entities[value])
+                    if isinstance(value, (list, dict)):
```

Comment:
Why is this validation at this point? It's already being validated in the beginning of the walk. 

It "saves" one iteration I assume?

Reply:
It's to be able to recursively "expand" dictionaries' values (which might be lists or dictionaries themselves).

Note that I'm checking against `value` not `node`